### PR TITLE
add Function templates for synchronizing CosmosDB and Redis Cache using Pub/Sub Trigger

### PR DESCRIPTION
**Pull request by royland123**
Added function templates for read through and write through database caching between Azure Cache for Redis and CosmosDB. 
Added a function template that writes Redis pub/sub messages to CosmosDB.

### CODE REVIEW

File: samples/PubSubSamples/WriteAroundSample/WriteAroundSample.cs [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+
+namespace Microsoft.Azure.WebJobs.Extensions.Redis.Samples
+{
+    public static class WriteAroundSample
+    {
+        public const string localhostSetting = "redisLocalhost";
+        public const string cosmosDbConnectionSetting = "CosmosDBConnection";
+
+        private static readonly IDatabaseAsync s_redisDb =
+            ConnectionMultiplexer.ConnectAsync("Rmac-Redis.redis.cache.windows.net:6380,password=t4sI8OxL5Ncwvo1fI6sxiJ6h5EzZtAAmmAzCaKZPSmE=,ssl=True,abortConnect=False,tiebreaker=").Result.GetDatabase();
```
Comment (royland123) [inline samples/PubSubSamples/WriteAroundSample/WriteAroundSample.cs:16]: Keys have been regenerated, sorry!

File: samples/PubSubSamples/RedisFunctionsDemo/WriteAroundSample.cs [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+namespace Microsoft.Azure.WebJobs.Extensions.Redis.Samples
+{
+    public static class WriteAroundSample
+    {
+        public const string localhostSetting = "redisLocalhost";
+        public const string cosmosDbConnectionSetting = "CosmosDBConnection";
+
+        private static readonly IDatabaseAsync s_redisDb =
+            ConnectionMultiplexer.ConnectAsync("<cache-name>.redis.cache.windows.net:6380,password=<access-key>,ssl=True,abortConnect=False,tiebreaker=").Result.GetDatabase();
```
Comment (dmonroym) [inline samples/PubSubSamples/RedisFunctionsDemo/WriteAroundSample.cs]: Using something like `ConnectionMultiplexer.ConnectAsync(configuration: ConfigurationManager.AppSettings["CacheConnection"].ToString()).Result.GetDatabase();` should get the connection from local.settings.json and that way you can avoid having this ever happening again

Comment (dmonroym) [reply samples/PubSubSamples/RedisFunctionsDemo/WriteAroundSample.cs]: Might actually be even better to use something like `Environment.GetEnvironmentVariable("key")` https://learn.microsoft.com/en-us/azure/azure-functions/functions-dotnet-class-library?tabs=v4%2Ccmd#environment-variables

Comment (royland123) [reply samples/PubSubSamples/RedisFunctionsDemo/WriteAroundSample.cs]: The latest commit uses `Environment.GetEnvironmentVariable("key")`

File: samples/PubSubSamples/WriteAroundSample/WriteAroundSample.cs (hunk 1 above)
Comment (dmonroym) [inline samples/PubSubSamples/WriteAroundSample/WriteAroundSample.cs:16]: !!!

File: samples/PubSubSamples/WriteBehindSample/WriteBehindSample.cs [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+    public static class WriteBehindSample
+    {
+        public const string localhostSetting = "redisLocalhost";
+        public const string cosmosDbConnectionSetting = "CosmosDBConnection";
+
+        private static readonly IDatabaseAsync s_redisDb =
+            ConnectionMultiplexer.ConnectAsync(Environment.GetEnvironmentVariable(localhostSetting)).Result.GetDatabase();
+
+
+        //write-behind caching: Write to Redis, then write to Cosmos asynchonously
```
Comment (dmonroym) [inline samples/PubSubSamples/WriteBehindSample/WriteBehindSample.cs:18]:
```suggestion
        //write-behind caching: Write to Redis, then write to Cosmos DB asynchronously
```

File: samples/PubSubSamples/WriteBehindSample/WriteBehindSample.cs [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+            RedisData redisData = new RedisData(
+                id: Guid.NewGuid().ToString(),
+                key: newKey,
+                value: await s_redisDb.StringGetAsync(newKey),
+                timestamp: DateTime.UtcNow
+                );
+
+            //write the record to cosmos
+            await cosmosOut.AddAsync(redisData);
+            logger.LogInformation($"key: \"{newKey}\" value: \"{redisData.value}\" addedd to cosmosdb container: \"{"ContainerId"}\" at id: \"{redisData.id}\"");
```
Comment (dmonroym) [inline samples/PubSubSamples/WriteBehindSample/WriteBehindSample.cs:38]:
```suggestion
            logger.LogInformation($"Key: \"{newKey}\" Value: \"{redisData.value}\" added to cosmos DB container: \"{"ContainerId"}\" at id: \"{redisData.id}\"");
```

File: samples/PubSubSamples/WriteThroughSample/WriteThroughSample.cs [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+           [CosmosDB(
+                databaseName: "DatabaseId",
+                containerName: "ContainerId",
+                Connection = cosmosDbConnectionSetting)]out dynamic redisData,
+           ILogger logger)
+        {
+          //get the redisDB synchronously
+            IDatabase redisDb = s_redisDb.Multiplexer.GetDatabase();
+
+            //assign the data from redis to a dyncmic object that will be written to cosmos
```
Comment (dmonroym) [inline samples/PubSubSamples/WriteThroughSample/WriteThroughSample.cs:30]:
```suggestion
            //assign the data from Redis to a dynamic object that will be written to cosmos
```

**Review by dmonroym — CHANGES_REQUESTED** [review]
Looks like there's some cleanup to be done. WriteAround files are there twice and should be deleted. Most of the comments in the ReadThroughSample.cs are applicable to the other functions as well

Adding some tests looks to be the next step to this.

File: samples/PubSubSamples/ReadThroughSample/ReadThroughSample.cs [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+
+        //keyspace notifications must be set to KEAm for this to trigger
+        //read-through caching: Read from Redis, if not found, read from Cosmos, then write to Redis
+        [FunctionName(nameof(ReadThroughAsync))]
+        public static async Task ReadThroughAsync(
+            [RedisPubSubTrigger(localhostSetting, "__keyevent@0__:keymiss")] string missedkey,
+            [CosmosDB(
+                databaseName: "DatabaseId",
+                containerName: "ContainerId",
+                Connection = cosmosDbConnectionSetting)]CosmosClient client,
```
[the remark this answers is not in the document]
Comment (royland123) [reply samples/PubSubSamples/ReadThroughSample/ReadThroughSample.cs:29]: `ConnectionStringSetting` has been deprecated and in versions 4.x+ it was replaced by `Connection`

File: samples/PubSubSamples/ReadThroughSample/ReadThroughSample.cs [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+using System.Threading.Tasks;
+
+namespace Microsoft.Azure.WebJobs.Extensions.Redis.Samples
+{
+    public static class ReadThroughSample
+    {
+        public const string localhostSetting = "redisLocalhost";
+        public const string cosmosDbConnectionSetting = "CosmosDBConnection";
+
+        private static readonly IDatabaseAsync s_redisDb =
```
[the remark this answers is not in the document]
Comment (royland123) [reply samples/PubSubSamples/ReadThroughSample/ReadThroughSample.cs]: Done!

File: samples/dotnet/Models/RedisData.cs [hunk 8]
```diff
@@ -0,0 +1,19 @@
+﻿using System;
```
Comment (royland123) [inline samples/dotnet/Models/RedisData.cs:1]: File can be removed

File: samples/dotnet/Microsoft.Azure.WebJobs.Extensions.Redis.Samples.csproj [hunk 9]
```diff
@@ -6,8 +6,11 @@
     <RepositoryUrl>https://github.com/Azure/azure-functions-redis-extension</RepositoryUrl>
   </PropertyGroup>
   <ItemGroup>
+    <PackageReference Include="Microsoft.Azure.Cosmos" Version="3.34.0" />
```
Comment (royland123) [inline samples/dotnet/Microsoft.Azure.WebJobs.Extensions.Redis.Samples.csproj:9]: unnecessary package references

**Comment by royland123** [comment]
local.settings.json files with dummy connection strings need to be added to each csproj

**Pull request closed without merging**